### PR TITLE
Update: add `ignoreDestructuring` option to `id-match` rule

### PR DESCRIPTION
**Pull request by tinymins**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**

add `ignoreDestructuring` option to `id-match` rule

**Is there anything you'd like reviewers to focus on?**

Nope.

Sorry that I do this code split job so late, cause I caught a cold and fever last week.

### CODE REVIEW

**Comment by tinymins** [comment]
@platinumazure I've recommitted, would you mind review again to see if there's any more problems should be fixed? Thanks a lot! :-)

**Comment by platinumazure** [comment]
Hi @tinymins, very sorry for losing track of this.

I'll champion this issue. (Will hopefully work with the ESLint team to accept this issue soon!)

I'll review this over the weekend!

**Comment by tinymins** [comment]
@platinumazure Thanks, with any problem in this PR, just point it out here, I'll check Github email everyday.

File: lib/rules/id-match.js [hunk 1]
```diff
@@ -106,36 +143,70 @@ module.exports = {
...
                         }
                     }
 
-                } else if (parent.type === "Property") {
+                /*
+                 * Properties have their own rules, and
+                 * AssignmentPattern nodes can be treated like Properties:
+                 * e.g.: const { no_camelcased = false } = bar;
+                 */
+                } else if (parent.type === "Property" || node.parent.type === "AssignmentPattern") {
```
Comment (Aladdin-ADD) [inline lib/rules/id-match.js]: > node.parent.type === "AssignmentPattern"

nits: is the `parent` means `node.parent`? I'd slightly prefer using `parent` for consistency.

File: docs/rules/id-match.md [hunk 2]
```diff
@@ -58,6 +58,9 @@ This rule has an object option:
 
 * `"properties": true` requires object properties to match the specified regular expression
 * `"onlyDeclarations": true` requires only `var`, `function`, and `class` declarations to match the specified regular expression
+* `"onlyDeclarations": false` requires all variable names to match the specified regular expression
```
Comment (Aladdin-ADD) [inline docs/rules/id-match.md]: nits: `"onlyDeclarations": false` (default) ?

File: lib/rules/id-match.js [hunk 3]
```diff
@@ -76,14 +110,17 @@ module.exports = {
...
         function report(node) {
-            context.report({
-                node,
-                message: "Identifier '{{name}}' does not match the pattern '{{pattern}}'.",
-                data: {
-                    name: node.name,
-                    pattern
-                }
-            });
+            if (reported.indexOf(node) < 0) {
```
[the remark this answers is not in the document]
Comment (tinymins) [reply lib/rules/id-match.js]: I agree with your idea, using map will absolutely faster than array. But I don't know if there is any straightforward way not reporting twice, most of those codes are inspired by this rule file: https://github.com/eslint/eslint/blob/066f7e07d4f722573c6dbe7f64b0765504595644/lib/rules/camelcase.js#L114

File: tests/lib/rules/id-match.js [hunk 4]
```diff
@@ -110,11 +110,42 @@ ruleTester.run("id-match", rule, {
             options: ["^[^_]+$"]
         },
         {
-            code: "var obj = {key: no_under}",
+            code: "var obj = {key: 'no_under'}",
```
[the remark this answers is not in the document]
Comment (tinymins) [reply tests/lib/rules/id-match.js]: My bad, I'll revert this change. I miss understand the test rule before.

File: lib/rules/id-match.js [hunk 5]
```diff
@@ -27,6 +27,12 @@ module.exports = {
                 properties: {
                     properties: {
                         type: "boolean"
+                    },
+                    onlyDeclarations: {
```
[the remark this answers is not in the document]
Comment (tinymins) [reply lib/rules/id-match.js]: 😅 Not expected, I made this mistake when separate commits.

File: docs/rules/id-match.md (hunk 2 above)
Comment (tinymins) [reply docs/rules/id-match.md]: I'll separate this code into another commit later.

File: lib/rules/id-match.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (tinymins) [reply lib/rules/id-match.js]: I've reviewed the code, this is expected. 😄 
`onlyDeclarations` already exist before my PR.

[the remark this answers is not in the document]
Comment (tinymins) [reply lib/rules/id-match.js]: This declaration may be forgotten before, I am just complete the missing part.

File: lib/rules/id-match.js (hunk 1 above)
Comment (tinymins) [reply lib/rules/id-match.js]: Yes, I'll make a change.

File: docs/rules/id-match.md (hunk 2 above)
Comment (tinymins) [reply docs/rules/id-match.md]: I don't know this. This feature was exist here for about two years. I've checked that `false` is its default value and I don't know why. Default value of `onlyDeclarations` became `false` is also for about two years.

File: lib/rules/id-match.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (tinymins) [reply lib/rules/id-match.js]: How to be a map? `node` is an object, and object cannot be key in object.

[the remark this answers is not in the document]
Comment (Aladdin-ADD) [reply lib/rules/id-match.js]: > there are important differences that make using a Map preferable in certain cases:
> The keys of an Object are Strings and Symbols, whereas they can be any value for a Map, including functions, objects, and any primitive.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#Objects_and_maps_compared

[the remark this answers is not in the document]
Comment (tinymins) [reply lib/rules/id-match.js]: Thanks a lot, I'll fix it. 😄

File: lib/rules/id-match.js [hunk 6]
```diff
@@ -106,36 +143,70 @@ module.exports = {
...
+                    }
+
+                    // never check properties or always ignore destructuring
+                    if (!properties || (ignoreDestructuring && isInsideObjectPattern(node))) {
                         return;
                     }
 
-                    if (shouldReport(effectiveParent, name)) {
+                    // don't check right hand side of AssignmentExpression to prevent duplicate warnings
+                    if (parent.right !== node && shouldReport(effectiveParent, name, true)) {
```
Comment (Aladdin-ADD) [inline lib/rules/id-match.js]: seems the 3rd  param `true` is not unused?

File: lib/rules/id-match.js [hunk 7]
```diff
@@ -106,36 +143,70 @@ module.exports = {
...
-                    }
-
-                    if (shouldReport(effectiveParent, name)) {
+                    // Report only if the local imported identifier is invalid
+                    if (parent.local && parent.local.name === node.name && isInvalid(name)) {
                         report(node);
                     }
+
+                // Report anything that is invalid that isn't a CallExpression
+                } else if (shouldReport(effectiveParent, name, false)) {
```
Comment (Aladdin-ADD) [inline lib/rules/id-match.js]: seems the 3rd  param `true` is not unused?

File: lib/rules/id-match.js [hunk 8]
```diff
@@ -106,36 +143,70 @@ module.exports = {
...
-                    if (shouldReport(effectiveParent, name)) {
+                    // don't check right hand side of AssignmentExpression to prevent duplicate warnings
+                    if (parent.right !== node && shouldReport(effectiveParent, name, true)) {
                         report(node);
                     }
 
-                } else {
-                    const isDeclaration = effectiveParent.type === "FunctionDeclaration" || effectiveParent.type === "VariableDeclarator";
+                // Check if it's an import specifier
+                } else if (["ImportSpecifier", "ImportNamespaceSpecifier", "ImportDefaultSpecifier"].indexOf(parent.type) >= 0) {
```
Comment (Aladdin-ADD) [inline lib/rules/id-match.js]: the array can be a set, and defined outside, just like `ALLOWED_PARENT_TYPES`?

File: lib/rules/id-match.js (hunk 6 above)
Comment (tinymins) [reply lib/rules/id-match.js]: My mistake, this param is not used for this commit.

**Pull request merged**